File: algorithms/core/misc/ListNode.java
package core.complex.misc;

import java.util.Iterator;


/**
 * Created by jhansen on 5/23/2015.
 */
public class ListNode<Item> implements Iterable<Item> {

    private ListNode<Item> next;
    private Item data;

    public ListNode( Item data, ListNode<Item> next ) {
        this.data = data;
        this.next = next;
    }

    public ListNode<Item> next() { return next; }
    public Item data() { return data; }

    public void next( ListNode<Item> next ) { this.next = next; }
    public void data( Item data ) { this.data = data; }

    public Iterator<Item> iterator() {
        return new NodeIterator<Item>(this);
    }

    private class NodeIterator<Item> implements Iterator<Item>
    {
        private ListNode<Item> current;

        public NodeIterator( ListNode<Item> first ) { current = first; }

        public boolean hasNext() { return current != null; }
        public Item next() {
            Item item = current.data;
            current = current.next;
            return item;
        }
    }
}

File: C#/algorithms/core/misc/ListNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace algorithms.core.misc {
    class ListNode<Item> : IEnumerable<Item> {
        private ListNode<Item> next;
        private Item data;

        public ListNode(Item data, ListNode<Item> next) {
            this.data = data;
            this.next = next;
        }

        public ListNode<Item> Next { get; set; }
        public Item Data { get; set; }

        public IEnumerator<Item> GetEnumerator() {
            ListNode<Item> current = this;
            while (current != null) {
                yield return current.data;
                current = current.next;
            }
        }

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() {
            return GetEnumerator();
        }

    }
}
